### PR TITLE
Rudimentary tracked-env control

### PR DESCRIPTION
**Pull request by pelson**
Adds the ability to:
- Automatically tag branch heads based on their commit date in the form `"tag_(env_name)_%Y_%m_%d"`
- Control next, current and previous tags with an interface.

### CODE REVIEW

File: conda_env_tracker/cli.py [hunk 1]
```diff
@@ -27,23 +27,29 @@
 
 def resolve_spec(spec_fh):
     """
-    Given an open file handle to an env.spec, return the package index,
-    the yaml interpretation of the handle, and the list of packages
-    which are resolved from the spec.
+    Given an open file handle to an env.spec, return a string containing
```
Comment (rhattersley) [inline conda_env_tracker/cli.py]: Mentioning that it returns a _list_ of strings would be helpful.

File: conda_env_tracker/cli.py [hunk 2]
```diff
@@ -27,23 +27,29 @@
...
+        pkgs.append('\t'.join([r['channel'], pkg[:-len('.tar.bz2')]])), 
+    return pkgs
 
 
 def build_manifest_branches(repo):
     for remote in repo.remotes:
         remote.fetch()
 
+    # Map channel to a conda index.
+    indices = {}
```
Comment (rhattersley) [inline conda_env_tracker/cli.py]: Unused.

File: conda_env_tracker/deploy.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+import os
+import time
+
+from git import Repo
+import conda.api
+import conda.fetch
+from conda_env_tracker.cli import tempdir, create_tracking_branches
+from conda_execute.lock import Locked
+
+manifest_branch_prefix = 'manifest/'
```
Comment (rhattersley) [inline conda_env_tracker/deploy.py]: It'd be good to share this with `conda_env_tracker/cli.py`.

File: conda_env_tracker/deploy.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+def deploy_tag(repo, tag_name, target):
+    tag = repo.tags[tag_name]
+    # Checkout the tag in a detached head form.
+    repo.head.reference = tag.commit
+    repo.head.reset(working_tree=True)
+
+    # Pull out the environment name from the form "env_<env_name>_2000_12_25".
+    env_name = tag_name.split('-')[1]
+    deployed_name = tag_name.split('-', 2)[2]
```
Comment (rhattersley) [inline conda_env_tracker/deploy.py:45]: Just a teaser ... but Python 3 syntax would be nice here:

``` python
_, env_name, deployed_name, *_ = tag_name.split('-')
```

Comment (rhattersley) [inline conda_env_tracker/deploy.py:45]: More seriously, your comment uses an underscore as the separator but the code uses a minus.

**Comment by pelson** [comment]
Thanks for the comments @rhattersley. I've been building on the work in the original PR and have ended up pushing it all into this PR. I've addressed your few comments, but they may well be lost in the noise.

I'm not pretending that this is beautiful, but I have got this running end-to-end every 5 minutes, and am _really_ happy with the result. I want to start separating the concerns, and building up smaller, less fragile tests (and as a result, improving the interfaces).

**Comment by rhattersley** [comment]
Top stuff! :+1: 

> I'm not pretending that this is beautiful

:+1: for brains over beauty right now.

> I want to start separating the concerns, and building up smaller, less fragile tests (and as a result, improving the interfaces).

I've mentioned the odd trivial thing, but deliberately held back from suggesting larger re-factorings. But now you've got a fair chunk of code :+1: for re-factoring as you go.

---

That's 3 :+1:s in one comment - it must be good! :wink: (Or should that read 4 :+1:s ... ABORT: recursion limit exceeded.)

**Pull request merged**